File: DataAlgo_forPy/Search_Sort/merge_sort.rb
# Implement Merge Sort

def merge_sort(array)
  return array if array.size <= 1

  mid = array.size / 2
  left = array[0...mid]
  right = array[mid..-1]

  merge_sort(left)
  merge_sort(right)

  i = 0
  j = 0
  k = 0

  while i < left.size && j < right.size
      if left[i] < right[j]
        array[k] = left[i]
        i += 1

      else
        array[k] = right[j]
        j += 1
      end

      k += 1
  end

  while i < left.size
    array[k] = left[i]
    i += 1
    k += 1
  end

  while j < right.size
    array[k] = right[j]
    j += 1
    k += 1
  end

  array
end


p merge_sort([6,4,1,2,4,6,2,10])

File: DataAlgo_forPy/Search_Sort/merge_sort.js
// Implement Merge Sort

const merge_sort = (array) => {
  if (array.length <= 1) return array

  const mid_point = Math.trunc(array.length / 2)

  const left = merge_sort(array.slice(0, mid_point))
  const right = merge_sort(array.slice(mid_point))

  return merge(left, right)
}

const merge = (left, right) => {
  const merged_array = []

  while (left.length > 0 && right.length > 0) {
    if (left[0] < right[0]) {
      merged_array.push(left.shift())
    } else if (left[0] === right[0]) {
      merged_array.push(left.shift())
    } else {
      merged_array.push(right.shift())
    }
  }

  return merged_array.concat(left, right)
}

console.log(merge_sort([6,4,1,2,4,6,2,10])) // [ 1, 2, 2, 4, 4, 6, 6, 10 ]
